File: Examples/ObjectiveCExample/ObjectiveCExample/BatchTestViewController.m
#import "BatchTestViewController.h"

#import "FixtureCatalog.h"
#import "GiftCatalog.h"

@import VAPPlayerKitObjC;

@interface BatchTestViewController () <VPKPlayerDelegate, VPKDynamicContentProvider>
@property (nonatomic, copy) NSArray<VAPFixture *> *fixtures;
@property (nonatomic, strong) VPKPlayerView *playerView;
@property (nonatomic, strong) UILabel *stateLabel;
@property (nonatomic, strong) UIProgressView *progressView;
@property (nonatomic, strong) UILabel *progressLabel;
@property (nonatomic, strong) UITextView *logView;
@property (nonatomic, assign) BOOL started;
@property (nonatomic, assign) BOOL cancelled;
@property (nonatomic, assign) BOOL currentDidStart;
@property (nonatomic, strong, nullable) NSError *currentFailure;
@property (nonatomic, strong, nullable) NSNumber *currentFinishReason;
@property (nonatomic, strong) NSMutableArray<NSString *> *lines;
@property (nonatomic, assign) NSUInteger runGeneration;
@end

@implementation BatchTestViewController

- (instancetype)initWithFixtures:(NSArray<VAPFixture *> *)fixtures {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _fixtures = [fixtures copy];
        _playerView = [[VPKPlayerView alloc] initWithFrame:CGRectZero];
        _stateLabel = [[UILabel alloc] init];
        _progressView = [[UIProgressView alloc] initWithProgressViewStyle:UIProgressViewStyleBar];
        _progressLabel = [[UILabel alloc] init];
        _logView = [[UITextView alloc] init];
        _lines = [NSMutableArray array];
    }
    return self;
}

- (void)dealloc {
    _cancelled = YES;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Device Batch Test";
    self.view.backgroundColor = [UIColor colorWithRed:0.03 green:0.04 blue:0.065 alpha:1];
    self.playerView.delegate = self;
    self.playerView.dynamicContentProvider = self;
    [self configureUI];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    if (self.started) {
        return;
    }
    self.started = YES;
    [self runAllFixtures];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    if (self.isMovingFromParentViewController) {
        self.cancelled = YES;
        self.runGeneration += 1;
        [self.playerView stop];
    }
}

- (void)configureUI {
    UIStackView *stack = [[UIStackView alloc] init];
    stack.translatesAutoresizingMaskIntoConstraints = NO;
    stack.axis = UILayoutConstraintAxisVertical;
    stack.spacing = 14;
    [self.view addSubview:stack];

    self.stateLabel.text = @"WAITING";
    self.stateLabel.font = [UIFont monospacedSystemFontOfSize:18 weight:UIFontWeightBold];
    self.stateLabel.textColor = [UIColor colorWithRed:0.48 green:0.84 blue:1 alpha:1];
    self.stateLabel.textAlignment = NSTextAlignmentCenter;
    self.stateLabel.accessibilityIdentifier = @"batch.state";
    self.progressView.progressTintColor = [UIColor colorWithRed:0.28 green:0.75 blue:0.57 alpha:1];
    self.progressLabel.font = [UIFont monospacedDigitSystemFontOfSize:12 weight:UIFontWeightMedium];
    self.progressLabel.textColor = UIColor.secondaryLabelColor;
    self.progressLabel.textAlignment = NSTextAlignmentCenter;
    self.progressLabel.accessibilityIdentifier = @"batch.progress";
    self.playerView.backgroundColor = [UIColor colorWithWhite:0.12 alpha:1];
    self.playerView.layer.cornerRadius = 18;
    self.playerView.clipsToBounds = YES;
    [self.playerView.heightAnchor constraintEqualToConstant:220].active = YES;
    self.logView.backgroundColor = [UIColor colorWithRed:0.02 green:0.025 blue:0.04 alpha:1];
    self.logView.textColor = [UIColor colorWithWhite:0.75 alpha:1];
    self.logView.font = [UIFont monospacedSystemFontOfSize:10 weight:UIFontWeightRegular];
    self.logView.editable = NO;
    self.logView.layer.cornerRadius = 14;
    self.logView.accessibilityIdentifier = @"batch.log";

    for (UIView *view in @[self.stateLabel, self.progressView, self.progressLabel, self.playerView, self.logView]) {
        [stack addArrangedSubview:view];
    }
    [NSLayoutConstraint activateConstraints:@[
        [stack.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:16],
        [stack.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:16],
        [stack.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-16],
        [stack.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-16]
    ]];
}

- (void)runAllFixtures {
    if (self.fixtures.count == 0) {
        self.stateLabel.text = @"BATCH FAILED";
        [self appendLine:@"No fixtures found"];
        return;
    }
    self.stateLabel.text = @"BATCH RUNNING";
    [self appendLine:[NSString stringWithFormat:@"Device: %@, iOS %@", UIDevice.currentDevice.name, UIDevice.currentDevice.systemVersion]];
    [self appendLine:[NSString stringWithFormat:@"Fixtures: %lu", (unsigned long)self.fixtures.count]];
    [self runFixtureAtIndex:0 failures:[NSMutableArray array] generation:self.runGeneration];
}

- (void)runFixtureAtIndex:(NSUInteger)index
                 failures:(NSMutableArray<NSString *> *)failures
               generation:(NSUInteger)generation {
    if (self.cancelled || self.runGeneration != generation) {
        return;
    }
    if (index >= self.fixtures.count) {
        [self.progressView setProgress:1 animated:YES];
        self.progressLabel.text = [NSString stringWithFormat:@"%lu / %lu", (unsigned long)self.fixtures.count, (unsigned long)self.fixtures.count];
        [self.playerView clear];
        if (failures.count == 0) {
            self.stateLabel.text = @"BATCH PASSED";
            [self appendLine:@"All fixture checks passed"];
        } else {
            self.stateLabel.text = @"BATCH FAILED";
            [self appendLine:[NSString stringWithFormat:@"Failures: %lu", (unsigned long)failures.count]];
            for (NSString *line in failures) {
                [self appendLine:line];
            }
        }
        return;
    }

    VAPFixture *fixture = self.fixtures[index];
    self.progressLabel.text = [NSString stringWithFormat:@"%lu / %lu · %@", (unsigned long)(index + 1), (unsigned long)self.fixtures.count, fixture.shortIdentifier];
    [self.progressView setProgress:(float)index / (float)self.fixtures.count animated:YES];
    self.currentDidStart = NO;
    self.currentFailure = nil;
    self.currentFinishReason = nil;
    [self.playerView clear];

    VPKPlaybackOptions *options = VPKPlaybackOptions.defaultOptions;
    options.loopCount = 0;
    options.clearsAfterFinish = NO;

    __weak typeof(self) weakSelf = self;
    [self.playerView prepareWithURL:fixture.url options:options completion:^(VPKAssetMetadata *metadata, NSError *error) {
        __strong typeof(weakSelf) self = weakSelf;
        if (self == nil || self.cancelled || self.runGeneration != generation) {
            return;
        }
        if (metadata == nil) {
            [self.playerView stop];
            if (fixture.looksLikeMedia) {
                [failures addObject:[NSString stringWithFormat:@"%@: %@", fixture.shortIdentifier, error.localizedDescription]];
                [self appendLine:[NSString stringWithFormat:@"FAIL %lu: %@", (unsigned long)(index + 1), error.localizedDescription]];
            } else {
                [self appendLine:[NSString stringWithFormat:@"PASS %lu: negative fixture rejected", (unsigned long)(index + 1)]];
            }
            [self runFixtureAtIndex:index + 1 failures:failures generation:generation];
            return;
        }
        if (!fixture.looksLikeMedia) {
            [failures addObject:[NSString stringWithFormat:@"%@: invalid fixture unexpectedly prepared", fixture.shortIdentifier]];
            [self appendLine:@"FAIL negative fixture accepted"];
            [self.playerView stop];
            [self runFixtureAtIndex:index + 1 failures:failures generation:generation];
            return;
        }
        if (metadata.frameCount <= 0 || metadata.duration <= 0 || metadata.canvasSize.width <= 0 || metadata.canvasSize.height <= 0) {
            [self.playerView stop];
            [failures addObject:[NSString stringWithFormat:@"%@: Metadata invariants failed.", fixture.shortIdentifier]];
            [self appendLine:[NSString stringWithFormat:@"FAIL %lu: Metadata invariants failed.", (unsigned long)(index + 1)]];
            [self runFixtureAtIndex:index + 1 failures:failures generation:generation];
            return;
        }

        [self.playerView playWithURL:fixture.url options:options];
        [self waitUntil:^BOOL {
            return self.currentDidStart;
        } timeout:3 generation:generation completion:^(NSError *waitError) {
            __strong typeof(weakSelf) self = weakSelf;
            if (self == nil || self.cancelled || self.runGeneration != generation) {
                return;
            }
            if (waitError != nil) {
                [self.playerView stop];
                [failures addObject:[NSString stringWithFormat:@"%@: %@", fixture.shortIdentifier, waitError.localizedDescription]];
                [self appendLine:[NSString stringWithFormat:@"FAIL %lu: %@", (unsigned long)(index + 1), waitError.localizedDescription]];
                [self runFixtureAtIndex:index + 1 failures:failures generation:generation];
                return;
            }
            NSTimeInterval keep = MIN(1, metadata.duration);
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(keep * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                __strong typeof(weakSelf) self = weakSelf;
                if (self == nil || self.cancelled || self.runGeneration != generation) {
                    return;
                }
                if (self.currentFailure != nil) {
                    [self.playerView stop];
                    [failures addObject:[NSString stringWithFormat:@"%@: %@", fixture.shortIdentifier, self.currentFailure.localizedDescription]];
                    [self appendLine:[NSString stringWithFormat:@"FAIL %lu: %@", (unsigned long)(index + 1), self.currentFailure.localizedDescription]];
                    [self runFixtureAtIndex:index + 1 failures:failures generation:generation];
                    return;
                }
                [self.playerView stop];
                [self waitUntil:^BOOL {
                    return self.currentFinishReason != nil && self.currentFinishReason.integerValue == VPKFinishReasonStopped;
                } timeout:3 generation:generation completion:^(NSError *stopError) {
                    __strong typeof(weakSelf) self = weakSelf;
                    if (self == nil || self.cancelled || self.runGeneration != generation) {
                        return;
                    }
                    if (stopError != nil || self.currentFinishReason.integerValue != VPKFinishReasonStopped) {
                        [failures addObject:[NSString stringWithFormat:@"%@: Stop did not emit the stopped terminal callback.", fixture.shortIdentifier]];
                        [self appendLine:[NSString stringWithFormat:@"FAIL %lu: Stop did not emit the stopped terminal callback.", (unsigned long)(index + 1)]];
                    } else {
                        [self appendLine:[NSString stringWithFormat:@"PASS %lu: %@, %ldf", (unsigned long)(index + 1), metadata.codec, (long)metadata.frameCount]];
                    }
                    [self runFixtureAtIndex:index + 1 failures:failures generation:generation];
                }];
            });
        }];
    }];
}

- (void)waitUntil:(BOOL (^)(void))predicate
          timeout:(NSTimeInterval)timeout
       generation:(NSUInteger)generation
       completion:(void (^)(NSError * _Nullable error))completion {
    NSTimeInterval deadline = NSProcessInfo.processInfo.systemUptime + timeout;
    [self pollUntil:predicate deadline:deadline generation:generation completion:completion];
}

- (void)pollUntil:(BOOL (^)(void))predicate
         deadline:(NSTimeInterval)deadline
       generation:(NSUInteger)generation
       completion:(void (^)(NSError * _Nullable error))completion {
    if (self.cancelled || self.runGeneration != generation) {
        return;
    }
    if (self.currentFailure != nil) {
        completion(self.currentFailure);
        return;
    }
    if (predicate()) {
        completion(nil);
        return;
    }
    if (NSProcessInfo.processInfo.systemUptime >= deadline) {
        completion([NSError errorWithDomain:@"ObjectiveCExample" code:1 userInfo:@{
            NSLocalizedDescriptionKey: @"Playback did not start before timeout."
        }]);
        return;
    }
    __weak typeof(self) weakSelf = self;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.05 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [weakSelf pollUntil:predicate deadline:deadline generation:generation completion:completion];
    });
}

- (void)appendLine:(NSString *)line {
    [self.lines addObject:line];
    self.logView.text = [self.lines componentsJoinedByString:@"\n"];
    if (self.logView.text.length > 0) {
        [self.logView scrollRangeToVisible:NSMakeRange(self.logView.text.length - 1, 1)];
    }
}

- (void)playerViewDidStart:(VPKPlayerView *)playerView {
    self.currentDidStart = YES;
}

- (void)playerView:(VPKPlayerView *)playerView didResolveMetadata:(VPKAssetMetadata *)metadata {
}

- (void)playerView:(VPKPlayerView *)playerView didFinishWithReason:(VPKFinishReason)reason {
    self.currentFinishReason = @(reason);
}

- (void)playerView:(VPKPlayerView *)playerView didFailWithError:(NSError *)error {
    self.currentFailure = error;
}

- (void)resolveTag:(NSString *)tag
            source:(VPKSourceMetadata *)source
        completion:(void (^)(UIImage * _Nullable, NSString * _Nullable, NSError * _Nullable))completion {
    [GiftCatalog resolveSource:source completion:completion];
}

@end

File: Examples/SwiftExample/SwiftExample/BatchTestViewController.swift
import UIKit
import VAPPlayerKit

/// 在真机上逐个验证所有 Bundle fixture：合法素材 prepare + 实际播放，非法素材必须明确失败。
final class BatchTestViewController: UIViewController, PlayerDelegate, DynamicContentProvider {
    private let fixtures: [VAPFixture]
    private let playerView = PlayerView()
    private let stateLabel = UILabel()
    private let progressView = UIProgressView(progressViewStyle: .bar)
    private let progressLabel = UILabel()
    private let logView = UITextView()
    private let renderProbe = RenderProbe()
    private var runTask: Task<Void, Never>?
    private var started = false
    private var currentDidStart = false
    private var currentFailure: Error?
    private var currentFinishReason: FinishReason?
    private var lines: [String] = []

    init(fixtures: [VAPFixture]) {
        self.fixtures = fixtures
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    deinit { runTask?.cancel() }

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Device Batch Test"
        view.backgroundColor = UIColor(red: 0.03, green: 0.04, blue: 0.065, alpha: 1)
        playerView.delegate = self
        playerView.dynamicContentProvider = self
        playerView.metricsSink = renderProbe
        configureUI()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        guard !started else { return }
        started = true
        runTask = Task { @MainActor [weak self] in await self?.runAllFixtures() }
    }

    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        if isMovingFromParent {
            runTask?.cancel()
            playerView.stop()
        }
    }

    private func configureUI() {
        let stack = UIStackView()
        stack.translatesAutoresizingMaskIntoConstraints = false
        stack.axis = .vertical
        stack.spacing = 14
        view.addSubview(stack)

        stateLabel.text = "WAITING"
        stateLabel.font = .monospacedSystemFont(ofSize: 18, weight: .bold)
        stateLabel.textColor = UIColor(red: 0.48, green: 0.84, blue: 1, alpha: 1)
        stateLabel.textAlignment = .center
        stateLabel.accessibilityIdentifier = "batch.state"
        progressView.progressTintColor = UIColor(red: 0.28, green: 0.75, blue: 0.57, alpha: 1)
        progressLabel.font = .monospacedDigitSystemFont(ofSize: 12, weight: .medium)
        progressLabel.textColor = .secondaryLabel
        progressLabel.textAlignment = .center
        progressLabel.accessibilityIdentifier = "batch.progress"
        playerView.backgroundColor = UIColor(white: 0.12, alpha: 1)
        playerView.layer.cornerRadius = 18
        playerView.clipsToBounds = true
        playerView.heightAnchor.constraint(equalToConstant: 220).isActive = true
        logView.backgroundColor = UIColor(red: 0.02, green: 0.025, blue: 0.04, alpha: 1)
        logView.textColor = UIColor(white: 0.75, alpha: 1)
        logView.font = .monospacedSystemFont(ofSize: 10, weight: .regular)
        logView.isEditable = false
        logView.layer.cornerRadius = 14
        logView.accessibilityIdentifier = "batch.log"

        [stateLabel, progressView, progressLabel, playerView, logView].forEach(stack.addArrangedSubview)
        NSLayoutConstraint.activate([
            stack.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 16),
            stack.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
            stack.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
            stack.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -16)
        ])
    }

    private func runAllFixtures() async {
        guard !fixtures.isEmpty else {
            stateLabel.text = "BATCH FAILED"
            append("No fixtures found")
            return
        }
        stateLabel.text = "BATCH RUNNING"
        append("Device: \(UIDevice.current.name), iOS \(UIDevice.current.systemVersion)")
        append("Fixtures: \(fixtures.count)")
        var failures: [String] = []

        for (index, fixture) in fixtures.enumerated() {
            guard !Task.isCancelled else { return }
            progressLabel.text = "\(index + 1) / \(fixtures.count) · \(fixture.shortIdentifier)"
            progressView.setProgress(Float(index) / Float(fixtures.count), animated: true)
            currentDidStart = false
            currentFailure = nil
            currentFinishReason = nil
            playerView.clear()
            let options = PlaybackOptions.defaultOptions
            options.loopCount = 0
            options.clearsAfterFinish = false

            do {
                let metadata = try await playerView.prepare(url: fixture.url, options: options)
                if !fixture.looksLikeMedia {
                    failures.append("\(fixture.shortIdentifier): invalid fixture unexpectedly prepared")
                    append("FAIL negative fixture accepted")
                    playerView.stop()
                    continue
                }
                guard metadata.frameCount > 0, metadata.duration > 0,
                      metadata.canvasSize.width > 0, metadata.canvasSize.height > 0 else {
                    throw BatchTestError.invalidMetadata
                }
                let initialRenderedCount = renderProbe.renderedFrameCount
                playerView.play(url: fixture.url, options: options)
                try await waitForPlaybackEvidence(after: initialRenderedCount, timeout: 3)
                try await keepPlaying(for: min(1, metadata.duration))
                playerView.stop()
                guard currentFinishReason == .stopped else { throw BatchTestError.missingStoppedCallback }
                append("PASS \(index + 1): \(metadata.codec), \(metadata.frameCount)f")
            } catch is CancellationError {
                return
            } catch {
                playerView.stop()
                if fixture.looksLikeMedia {
                    failures.append("\(fixture.shortIdentifier): \(error.localizedDescription)")
                    append("FAIL \(index + 1): \(error.localizedDescription)")
                } else {
                    append("PASS \(index + 1): negative fixture rejected")
                }
            }
        }

        progressView.setProgress(1, animated: true)
        progressLabel.text = "\(fixtures.count) / \(fixtures.count)"
        playerView.clear()
        if failures.isEmpty {
            stateLabel.text = "BATCH PASSED"
            append("All fixture checks passed")
        } else {
            stateLabel.text = "BATCH FAILED"
            append("Failures: \(failures.count)")
            failures.forEach { append($0) }
        }
    }

    private func append(_ line: String) {
        lines.append(line)
        logView.text = lines.joined(separator: "\n")
        if !logView.text.isEmpty {
            logView.scrollRangeToVisible(NSRange(location: logView.text.utf16.count - 1, length: 1))
        }
    }

    private func waitForPlaybackEvidence(after renderedCount: Int, timeout: TimeInterval) async throws {
        let deadline = ProcessInfo.processInfo.systemUptime + timeout
        while ProcessInfo.processInfo.systemUptime < deadline {
            if let currentFailure { throw currentFailure }
            if currentDidStart, renderProbe.renderedFrameCount > renderedCount { return }
            try await Task.sleep(nanoseconds: 50_000_000)
        }
        throw BatchTestError.noRenderedFrame
    }

    private func keepPlaying(for duration: TimeInterval) async throws {
        let deadline = ProcessInfo.processInfo.systemUptime + max(0, duration)
        while ProcessInfo.processInfo.systemUptime < deadline {
            if let currentFailure { throw currentFailure }
            try await Task.sleep(nanoseconds: 50_000_000)
        }
    }

    func playerDidStart(_ player: PlayerView) { currentDidStart = true }
    func player(_ player: PlayerView, didUpdate metadata: AssetMetadata) {}
    func playerDidFinish(_ player: PlayerView, reason: FinishReason) { currentFinishReason = reason }
    func player(_ player: PlayerView, didFail error: Error) { currentFailure = error }

    func resolve(
        tag: String,
        source: SourceMetadata,
        completion: @escaping (DynamicContent?, Error?) -> Void
    ) {
        if tag.localizedCaseInsensitiveContains("text") || tag.localizedCaseInsensitiveContains("name") {
            completion(.textReplacement("VAP Swift"), nil)
            return
        }
        let image = UIGraphicsImageRenderer(size: source.slotSize).image { context in
            UIColor.systemBlue.setFill()
            context.cgContext.fill(CGRect(origin: .zero, size: source.slotSize))
        }
        completion(.image(image), nil)
    }
}

private enum BatchTestError: LocalizedError {
    case invalidMetadata
    case noRenderedFrame
    case missingStoppedCallback

    var errorDescription: String? {
        switch self {
        case .invalidMetadata: return "Metadata invariants failed."
        case .noRenderedFrame: return "Playback did not start and render a frame before timeout."
        case .missingStoppedCallback: return "Stop did not emit the stopped terminal callback."
        }
    }
}

private final class RenderProbe: MetricsSink {
    private let lock = NSLock()
    private var renderedFrames = 0

    var renderedFrameCount: Int {
        lock.lock()
        defer { lock.unlock() }
        return renderedFrames
    }

    func record(_ event: MetricsEvent) {
        guard case .renderedFrame = event else { return }
        lock.lock()
        renderedFrames += 1
        lock.unlock()
    }
}
